src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs

```diff
@@ -258,23 +268,46 @@ public static DirectoryInfo CreateDirectory(this DirectorySecurity directorySecu
 
         // In the context of a FileStream, the only ACCESS_MASK ACE rights we care about are reading/writing data and the generic read/write rights.
         // See: https://docs.microsoft.com/en-us/windows/win32/secauthz/access-mask
-        private static FileAccess GetFileStreamFileAccess(FileSystemRights rights)
+        private static FileAccess GetFileAccessFromRights(FileSystemRights rights)
         {
             FileAccess access = 0;
-            if ((rights & FileSystemRights.ReadData) != 0 || ((int)rights & Interop.Kernel32.GenericOperations.GENERIC_READ) != 0)
+
+            if ((rights & FileSystemRights.FullControl) != 0 ||
+                (rights & FileSystemRights.Modify) != 0)
+            {
+                return FileAccess.ReadWrite;
+            }
+
+            if ((rights & FileSystemRights.ReadData) != 0 || // Same as ListDirectory
+                (rights & FileSystemRights.ReadExtendedAttributes) != 0 ||
+                (rights & FileSystemRights.ExecuteFile) != 0 || // Same as Traverse
+                (rights & FileSystemRights.ReadAttributes) != 0 ||
+                (rights & FileSystemRights.ReadPermissions) != 0 ||
+                (rights & FileSystemRights.TakeOwnership) != 0 ||
+                (rights & FileSystemRights.Synchronize) != 0 ||
+                ((int)rights & Interop.Kernel32.GenericOperations.GENERIC_READ) != 0)
             {
                 access = FileAccess.Read;
             }
 
-            if ((rights & FileSystemRights.WriteData) != 0 || ((int)rights & Interop.Kernel32.GenericOperations.GENERIC_WRITE) != 0)
+            if ((rights & FileSystemRights.AppendData) != 0 || // Same as CreateDirectories
+                (rights & FileSystemRights.ChangePermissions) != 0 ||
+                (rights & FileSystemRights.Delete) != 0 ||
+                (rights & FileSystemRights.DeleteSubdirectoriesAndFiles) != 0 ||
+                (rights & FileSystemRights.WriteAttributes) != 0 ||
+                (rights & FileSystemRights.WriteData) != 0 || // Same as CreateFiles
+                (rights & FileSystemRights.WriteExtendedAttributes) != 0 ||
+                ((int)rights & Interop.Kernel32.GenericOperations.GENERIC_WRITE) != 0)
             {
-                access = access == FileAccess.Read ? FileAccess.ReadWrite : FileAccess.Write;
+                access |= FileAccess.Write;
             }
 
+            Debug.Assert(access != 0);
+
             return access;
         }
 
-        private static unsafe SafeFileHandle CreateFileHandle(string fullPath, FileMode mode, FileSystemRights rights, FileShare share, FileOptions options, FileSecurity security)
+        private static unsafe SafeFileHandle CreateFileHandle(string fullPath, FileMode mode, FileSystemRights rights, FileShare share, FileOptions options, FileSecurity? security)
```

Comment:
this logic is very similar to https://github.com/dotnet/runtime/blob/39ece73bd08b1bcea0b9447cf34690bb1b41ba0a/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs#L52-L102

would it be possible to unify that and keep it in a single `.cs`. file that would be referenced by both projects?

Reply:
Sounds reasonable. Do you mind if I try it in a separate PR?

Reply:
> Do you mind if I try it in a separate PR?

I don't mind.

Reply:
https://github.com/dotnet/runtime/issues/64017